### PR TITLE
update source-s3 to use airbyte-cdk version with protocol pydantic v2

### PR DESCRIPTION
**Pull request by erohmensing**
## What

* Speed up source s3

## How

* Upgrade to CDK version 2.0
  * Update the pydantic imports according to the migration guide 
  * Fix a change in the spec introduced in cdk version 1.3.1

## User Impact

Faster syncs!

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by erohmensing** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39573?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#39573** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39573?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#39524** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39524?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @erohmensing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-integrations/connectors/source-s3/integration_tests/cloud_spec.json [hunk 1]
```diff
@@ -285,7 +285,7 @@
                   "required": ["filetype"]
                 },
                 {
-                  "title": "Document File Type Format (Experimental)",
+                  "title": "Unstructured Document Format",
```
Comment (erohmensing) [inline airbyte-integrations/connectors/source-s3/integration_tests/cloud_spec.json:288]: These changes are unrelated to 2.0 - they come from upgrading through CDK version 1.3.1.

File: docs/integrations/sources/s3.md [hunk 2]
```diff
@@ -345,6 +345,7 @@ This connector utilizes the open source [Unstructured](https://unstructured-io.g
 
 | Version | Date       | Pull Request                                                                                                    | Subject                                                                                                              |
 | :------ | :--------- | :-------------------------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------- |
+| 4.6.0 | 2024-06-21 | [39573](https://github.com/airbytehq/airbyte/pull/39573) | Update to Airbyte CDK 2.0.0 |
```
Comment (maxi297) [inline docs/integrations/sources/s3.md:348]: nit: the date is in the past

nit: I would add that the main goal is to improve performance

File: airbyte-integrations/connectors/source-s3/poetry.lock [hunk 3]
```diff
@@ -865,13 +876,13 @@ jsonpointer = ">=1.9"
 
 [[package]]
 name = "jsonpointer"
-version = "2.4"
+version = "3.0.0"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-s3/poetry.lock:879]: This major change not related to pydantic got me worried a bit. The path to this dependency is: airbyte-cdk -> langchain-core -> jsonpatch -> jsonpointer

Given we don't use langchain-core in source-s3, this should be fine

**Review by maxi297 — APPROVED** [review]
LGTM! Have we set up a dashboard to track the performance and see the improvement? For Concurrent CDK, we created [this one](https://airbyte.metabaseapp.com/dashboard/2085-concurrent-cdk?date_filter=past30days) that is now broken because we don't sync the connection anymore. We also only relied on the duration because we didn't have the proper metric to calculate the source speed (which we just ignore as we could manually see that all the data was there)

File: docs/integrations/sources/s3.md (hunk 2 above)
Comment (erohmensing) [reply docs/integrations/sources/s3.md:348]: Good callouts! Will update

File: airbyte-integrations/connectors/source-s3/poetry.lock (hunk 3 above)
Comment (erohmensing) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: this is actually a good question - im not sure why langchain is being installed in any case 🤔

Comment (maxi297) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: I also don't know how it works. It is [part of the optional dependencies](https://github.com/airbytehq/airbyte/blob/48aa409d04047bc82e63d43e0271a95c7b3a9a1d/airbyte-cdk/python/pyproject.toml#L50). Is poetry installing optional things as well? @alafanechere maybe? 

In any case, I don't think this is a blocker as we should not use this as part of source-s3

Comment (erohmensing) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: Cool, I don't think so either, I was also looking for augustin for help though :D

Comment (alafanechere) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: I do think the default poetry install command installs all declared extras

Comment (erohmensing) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: We specify extras that we want to install, but we install file-based and not vector-db-based

Comment (erohmensing) [reply airbyte-integrations/connectors/source-s3/poetry.lock:879]: maybe its the `--no-dev` that we need to add to not get it to install everything

**Comment by erohmensing** [comment]
@maxi297 we had a meeting yesterday about tracking performance, and have a lot of benchmarks to set up. For this one, will make sure to run a sync pinned on the older version before getting it up to the newer version for comparison 👍
